packages/atomic/src/utils/initialization-utils.tsx

```diff
@@ -1,53 +1,132 @@
-import {Engine} from '@coveo/headless';
+import {Controller, Engine} from '@coveo/headless';
 import {ComponentInterface, getElement, h} from '@stencil/core';
 import {i18n} from 'i18next';
 
-export interface InterfaceContext {
+/**
+ * Bindings passed from the `AtomicSearchInterface` to its children components.
+ */
+export interface Bindings {
+  /**
+   * Headless Engine instance.
+   */
   engine: Engine;
+  /**
+   * i18n instance, for localization.
+   */
   i18n: i18n;
 }
 
-export type InitializeEventHandler = (context: InterfaceContext) => void;
+export type InitializeEventHandler = (bindings: Bindings) => void;
 export type InitializeEvent = CustomEvent<InitializeEventHandler>;
 
-export class InitializationError extends Error {
+export class MissingInterfaceParentError extends Error {
   constructor(elementName: string) {
     super(
       `The "${elementName}" element must be the child of a configured "atomic-search-interface" element.`
     );
-    this.name = 'InitializationError';
   }
 }
 
-export function Initialization() {
-  return (component: ComponentInterface, initializeMethod: string) => {
+/**
+ * Necessary interface an Atomic Component must have to initialize itself correctly.
+ */
+export interface AtomicComponentInterface extends ComponentInterface {
+  /**
+   * Bindings passed from the `AtomicSearchInterface` to its children components.
+   */
+  bindings: Bindings;
+  /**
+   * Record of methods earch returning an i18n localized string.
+   */
+  strings?: Record<string, () => string>;
+  /**
+   * Headless Controller instance associated with the Atomic Component.
+   */
+  controller?: Controller;
+  /**
+   * Headless Controller's state.
+   */
+  controllerState?: unknown;
+  /**
+   * Callback for when the subscribe method is called and the controller's state is updated.
+   */
+  onControllerStateUpdate?: () => void;
```

Comment:
I'm on the fence with this recent addition. I think placing too much magic inside the Initialization decorator makes Atomic harder to understand.

We are assuming that there is one controller for every component. This may be the case for the components we build. However, if someone wants to build a custom component, they might want to use more than one controller together, or actions directly with no controller. Needing to keep my instance in a property called `controller`, and read it's state from a property called `controllerState` feels restrictive, and something I need to know.

I would suggest keeping the initialization decorator focused on bindings. The component sub/upsub/state update logic could be managed by a separate decorator, or just repeating it in the components.

This would avoid options like `resubscribeControllerOnConnectedCallback`. If someone needs to use a certain lifecycle method, they can do so without needing to edit the decorator. They can also use more meaningful names for their properties.

Those are my thoughts, I leave the final decision to you,

Reply:
I think its a pretty narrow edge case where you would actually want to use two controllers inside one component where composition doesn't make more sense. And if they just don't provide a controller they can just ignore `controllerState` and initialize the controller using the engine inside the bindings variable?

Personally I feel like for the vast majority of use cases having Atomic handle this for you is cleaner. With that said you're right that it does create more Atomic specific knowledge.

Reply:
As long as we keep it so "binding" is the only actually required thing in there, I think it's fine.

For the other props, they are optional, and they help cover a vast majority of use case/frequent pattern.

And if you have something different, you don't have to use @controllerState for example, or any other,

You can go free style ;)

As for separating them in different decorator that you can **choose** to use or not, or one decorator for which you can **choose** which prop to use or not, it's mostly a stylistic change, and does not seem like a fundamental one.

And if we see we never use the decorator for what it was designed to do, as in we see everybody going "free style", we revisit.